Add overload check and recovery to realtime queue

lordnull opened the pull request.

lordnull committed:
Added tests for start with overload options

Also renamed rtq test module so it will actually be run duing a
rebar eunit (or make test).

lordnull committed:
Overload and recover implemented.

lordnull committed:
Changed defaults for overload

Also commented on a 'trick' used to conditionally enable lager when
running tests.

lordnull committed:
able to flip overload on drop report as well

Vagabond commented on src/riak_repl2_rtq.erl:
```diff
@@ -73,7 +88,27 @@
+start_link(Options) ->
+    case ets:info(?overload_ets) of
+        undefined ->
+            ets:new(?overload_ets, [named_table, public, {read_concurrency, true}]),
```
Should this not be a protected table?

Vagabond replied:
Also, in start_link we are running in the context of the supervisor, not the RTQ?

lordnull replied:
Commented earlier, but github ate it.

There was discussion about the actual implementation detail of counting the drops. One of the options was putting the counter in the ets, and having the overload just read and reset it occasionally. The only way to do this is by making the ets public. Putting it under the supervisor means drop count doesn't get reset if the rtq would die.

Vagabond commented on src/riak_repl2_rtq.erl:
```diff
@@ -73,7 +88,27 @@
-    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+    Overload = app_helper:get_env(riak_repl, rtq_overload_threshold, ?DEFAULT_OVERLOAD),
+    Recover = app_helper:get_env(riak_repl, rtq_overload_recover, ?DEFAULT_RECOVER),
+    Opts = [{overload_threshold, Overload}, {overload_recover, Recover}],
```
Why default overload and recover here, but also in init()?

lordnull replied:
It's defaulted here to ensure it's sent to start_link/1 based on what is in app.config, but still with a valid value. Having it in init/1 means start_link([]) can be used for testing.

Vagabond commented:
I think you forgot to add the overload_counter module.

Vagabond commented on test/riak_repl2_rtq_tests.erl:
```diff
@@ -0,0 +1,196 @@
+            [riak_repl2_rtq:push(1, term_to_binary([<<"object">>])) || _ <- lists:seq(1,5)],
+            % msq queue = 7, drops = 5
+            unblock_rtq_pull(),
+            timer:sleep(1200),
```
If possible, don't use timers likethis, because the test will fail on slower machines. Can we do something like send the RTQ a status request so we can serialize on the mailbox? Even a sys:get_status may do the trick...

lordnull replied:
The timer isn't waiting for rtq, though, it's waiting for a message from rtq_overload_counter to be sent to the rtq, which itself is triggered by a timer. I agree it's better not to use a timer here, but I'm not sure how I would do it the 'right' way.

lordnull committed:
Put the overload counter in the supervisor tree.

As well as in the .app file.

lordnull committed:
Added missing overload module.

lordnull committed:
going into overload now sets rt_dirty.

rt_source_errors in the riak_repl_stats sets rt_dirty, so it's an
indirect test.

lordnull committed:
overload_drops added to rtq status.

Vagabond commented on src/riak_repl2_rtq.erl:
```diff
@@ -340,6 +402,25 @@ terminate(Reason, #state{cs = Cs}) ->
+maybe_flip_overload(State) ->
+    #state{overloaded = Overloaded, overload = Overload, recover = Recover} = State,
+    {message_queue_len, MsgQLen} = erlang:process_info(self(), message_queue_len),
+    lager:debug("checking overload: ~p", [MsgQLen]),
```
debug messages aren't entirely free, so leaving one of doubtful value in the hot path should be avoided.

Vagabond commented:
I'm generally +1 on this from a code perspective (nuking that lager:debug would be nice). I have not run the code at all, so maybe @bowrocker should also +1 it from a testing perspective.

lordnull committed:
Removed a debug of dubius value.

bowrocker commented:
+1 from running lots and lots of  tests.
